lib/OrdinaryDiffEqDifferentiation/src/alg_utils.jl

```diff
@@ -41,58 +41,113 @@ end
+    SciMLBase.@set autodiff_alg.dense_ad = prepare_ADType(ADTypes.dense_ad(autodiff_alg), prob, u0, p, standardtag)
+end
+
+function prepare_ADType(autodiff_alg::AutoForwardDiff, prob, u0, p, standardtag)
+    tag = if standardtag
+        ForwardDiff.Tag(OrdinaryDiffEqTag(), eltype(u0))
+    else
+        nothing
     end
 
+    T = eltype(u0)
+
+    fwd_cs = OrdinaryDiffEqCore._get_fwd_chunksize_int(autodiff_alg)
+
+    cs = fwd_cs == 0 ? nothing : fwd_cs
+
+    if ((prob.f isa ODEFunction &&
+      prob.f.f isa FunctionWrappersWrappers.FunctionWrappersWrapper) ||
+     (isbitstype(T) && sizeof(T) > 24)) && (cs == 0 || isnothing(cs))
+        return AutoForwardDiff{1}(tag)
+    else 
+        return AutoForwardDiff{cs}(tag)
+    end
+end
+
+function prepare_ADType(alg::AutoFiniteDiff, prob, u0, p, standardtag)
     # If the autodiff alg is AutoFiniteDiff, prob.f.f isa FunctionWrappersWrapper,
     # and fdtype is complex, fdtype needs to change to something not complex
-    if alg_autodiff(alg) isa AutoFiniteDiff
-        if alg_difftype(alg) == Val{:complex} && (prob.f isa ODEFunction &&
-            prob.f.f isa FunctionWrappersWrappers.FunctionWrappersWrapper)
-            @warn "AutoFiniteDiff fdtype complex is not compatible with this function"
-            return remake(alg, autodiff = AutoFiniteDiff(fdtype = Val{:forward}()))
-        end
-        return alg
+    if alg.fdtype == Val{:complex}() && (prob.f isa ODEFunction && prob.f.f isa FunctionWrappersWrappers.FunctionWrappersWrapper)
+         @warn "AutoFiniteDiff fdtype complex is not compatible with this function"
+         return AutoFiniteDiff(fdtype = Val{:forward}())
     end
+    return alg
+end
+
+function prepare_user_sparsity(ad_alg, prob)
+    jac_prototype = prob.f.jac_prototype
+    sparsity = prob.f.sparsity
+
+    if !isnothing(sparsity) && !(ad_alg isa AutoSparse)
+        if sparsity isa SparseMatrixCSC
+            if prob.f.mass_matrix isa UniformScaling
+                idxs = diagind(sparsity)
+                @. @view(sparsity[idxs]) = 1
+
+                if !isnothing(jac_prototype)
+                    @. @view(jac_prototype[idxs]) = 1
+                end
+            else
+                idxs = findall(!iszero, prob.f.mass_matrix)
+                for idx in idxs
+                    sparsity[idx] = prob.f.mass_matrix[idx]
+                end
 
-    L = StaticArrayInterface.known_length(typeof(u0))
-    if L === nothing # dynamic sized
-        # If chunksize is zero, pick chunksize right at the start of solve and
-        # then do function barrier to infer the full solve
-        x = if prob.f.colorvec === nothing
-            length(u0)
-        else
-            maximum(prob.f.colorvec)
+                if !isnothing(jac_prototype)
+                    for idx in idxs
+                        jac_prototype[idx] = f.mass_matrix[idx]
+                    end
+                end
+            end
```

Comment:
@jClugstor what is the rationale behind these lines? I didn't catch it at first but I think they are causing bugs downstream. You cannot change the sparsity pattern without making sure that the detector will also return that same pattern.

Reply:
The jac function is slightly different from the f, so that's required.

Reply:
Let's move the discussion to #2653 because this needs an urgent fix